File: tests/build_disk_index.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

#include <omp.h>
#include <boost/program_options.hpp>

#include "aux_utils.h"
#include "index.h"
#include "math_utils.h"
#include "partition_and_pq.h"
#include "utils.h"

namespace po = boost::program_options;

int main(int argc, char** argv) {
  std::string data_type, dist_fn, data_path, index_path_prefix;
  unsigned    num_threads, R, L, disk_PQ;
  float       B, M;
  bool        append_reorder_data = false;
  bool        use_opq = false;

  po::options_description desc{"Arguments"};
  try {
    desc.add_options()("help,h", "Print information on arguments");
    desc.add_options()("data_type",
                       po::value<std::string>(&data_type)->required(),
                       "data type <int8/uint8/float>");
    desc.add_options()("dist_fn", po::value<std::string>(&dist_fn)->required(),
                       "distance function <l2/mips>");
    desc.add_options()("data_path",
                       po::value<std::string>(&data_path)->required(),
                       "Input data file in bin format");
    desc.add_options()("index_path_prefix",
                       po::value<std::string>(&index_path_prefix)->required(),
                       "Path prefix for saving index file components");
    desc.add_options()("max_degree,R",
                       po::value<uint32_t>(&R)->default_value(64),
                       "Maximum graph degree");
    desc.add_options()(
        "Lbuild,L", po::value<uint32_t>(&L)->default_value(100),
        "Build complexity, higher value results in better graphs");
    desc.add_options()("search_DRAM_budget,B", po::value<float>(&B)->required(),
                       "DRAM budget in GB for searching the index to set the "
                       "compressed level for data while search happens");
    desc.add_options()("build_DRAM_budget,M", po::value<float>(&M)->required(),
                       "DRAM budget in GB for building the index");
    desc.add_options()(
        "num_threads,T",
        po::value<uint32_t>(&num_threads)->default_value(omp_get_num_procs()),
        "Number of threads used for building index (defaults to "
        "omp_get_num_procs())");
    desc.add_options()("PQ_disk_bytes",
                       po::value<uint32_t>(&disk_PQ)->default_value(0),
                       "Number of bytes to which vectors should be compressed "
                       "on SSD; 0 for no compression");
    desc.add_options()("append_reorder_data",
                       po::bool_switch()->default_value(false),
                       "Include full precision data in the index. Use only in "
                       "conjuction with compressed data on SSD.");

    desc.add_options()("use_opq", po::bool_switch()->default_value(false),
                       "Use Optimized Product Quantization (OPQ).");

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    if (vm.count("help")) {
      std::cout << desc;
      return 0;
    }
    po::notify(vm);
    if (vm["append_reorder_data"].as<bool>())
      append_reorder_data = true;
    if (vm["use_opq"].as<bool>())
      use_opq = true;
  } catch (const std::exception& ex) {
    std::cerr << ex.what() << '\n';
    return -1;
  }

  diskann::Metric metric;
  if (dist_fn == std::string("l2"))
    metric = diskann::Metric::L2;
  else if (dist_fn == std::string("mips"))
    metric = diskann::Metric::INNER_PRODUCT;
  else {
    std::cout << "Error. Only l2 and mips distance functions are supported"
              << std::endl;
    return -1;
  }

  if (append_reorder_data) {
    if (disk_PQ == 0) {
      std::cout << "Error: It is not necessary to append data for reordering "
                   "when vectors are not compressed on disk."
                << std::endl;
      return -1;
    }
    if (data_type != std::string("float")) {
      std::cout << "Error: Appending data for reordering currently only "
                   "supported for float data type."
                << std::endl;
      return -1;
    }
  }

  std::string params = std::string(std::to_string(R)) + " " +
                       std::string(std::to_string(L)) + " " +
                       std::string(std::to_string(B)) + " " +
                       std::string(std::to_string(M)) + " " +
                       std::string(std::to_string(num_threads)) + " " +
                       std::string(std::to_string(disk_PQ)) + " " +
                       std::string(std::to_string(append_reorder_data));

  try {
    if (data_type == std::string("int8"))
      return diskann::build_disk_index<int8_t>(data_path.c_str(),
                                               index_path_prefix.c_str(),
                                               params.c_str(), metric, use_opq);
    else if (data_type == std::string("uint8"))
      return diskann::build_disk_index<uint8_t>(
          data_path.c_str(), index_path_prefix.c_str(), params.c_str(), metric,
          use_opq);
    else if (data_type == std::string("float"))
      return diskann::build_disk_index<float>(data_path.c_str(),
                                              index_path_prefix.c_str(),
                                              params.c_str(), metric, use_opq);
    else {
      diskann::cerr << "Error. Unsupported data type" << std::endl;
      return -1;
    }
  } catch (const std::exception& e) {
    std::cout << std::string(e.what()) << std::endl;
    diskann::cerr << "Index build failed." << std::endl;
    return -1;
  }
}

File: diskann-tools/src/bin/build_disk_index.rs
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT license.
 */

use clap::Parser;
use diskann::{graph::config::defaults::ALPHA, ANNError};
use diskann_disk::QuantizationType;
use diskann_providers::storage::FileStorageProvider;
use diskann_tools::utils::{
    build_disk_index, get_num_threads, init_subscriber, BuildDiskIndexParameters, CMDToolError,
    DataType, DimensionValues, GraphDataF32Vector, GraphDataHalfVector, GraphDataInt8Vector,
    GraphDataMinMaxVector, GraphDataU8Vector,
};
use diskann_vector::distance::Metric;

fn main() -> Result<(), CMDToolError> {
    init_subscriber();

    let args: BuildDiskIndexArgs = BuildDiskIndexArgs::parse();

    let threads = get_num_threads(args.num_threads);

    println!(
        "Starting index build with R: {}  Lbuild: {}  alpha: {}  #threads: {} num_of_pq_chunks: {} build_DRAM_budget: {}",
        args.max_degree, args.l_build, ALPHA, threads, args.num_of_pq_chunks, args.build_dram_budget
    );

    let parameters = BuildDiskIndexParameters {
        metric: args.dist_fn,
        data_path: &args.data_path,
        r: args.max_degree,
        l: args.l_build,
        index_path_prefix: &args.index_path_prefix,
        num_threads: threads,
        index_build_ram_limit_gb: args.build_dram_budget,
        build_quantization_type: args.build_quantization_type,
        num_of_pq_chunks: args.num_of_pq_chunks,
        chunking_parameters: None,
        dim_values: DimensionValues::new(args.dim, args.dim),
    };

    let storage_provider = FileStorageProvider;

    let err = if args.use_minmax {
        if args.full_dim.is_none() {
            return Err(ANNError::log_index_config_error(
                format!("Full dim : {:?}", args.full_dim),
                "full_dim cannot be None for minmax based build".to_string(),
            )
            .into());
        }
        let parameters = BuildDiskIndexParameters {
            dim_values: DimensionValues::new(args.dim, args.full_dim.unwrap()),
            ..parameters
        };
        build_disk_index::<GraphDataMinMaxVector, FileStorageProvider>(
            &storage_provider,
            parameters,
        )
    } else {
        match args.data_type {
            DataType::Int8 => build_disk_index::<GraphDataInt8Vector, FileStorageProvider>(
                &storage_provider,
                parameters,
            ),
            DataType::Uint8 => build_disk_index::<GraphDataU8Vector, FileStorageProvider>(
                &storage_provider,
                parameters,
            ),
            DataType::Float => build_disk_index::<GraphDataF32Vector, FileStorageProvider>(
                &storage_provider,
                parameters,
            ),
            DataType::Fp16 => build_disk_index::<GraphDataHalfVector, FileStorageProvider>(
                &storage_provider,
                parameters,
            ),
        }
    };

    match err {
        Ok(_) => {
            println!("Index build completed successfully");
            Ok(())
        }
        Err(err) => {
            tracing::error!("Index build failed - see diagnostic");
            Err(err.into())
        }
    }
}

#[derive(Debug, Parser)]
struct BuildDiskIndexArgs {
    /// data type <int8/uint8/float / fp16> (required)
    #[arg(long = "data_type", default_value = "float")]
    pub data_type: DataType,

    /// Should minmax vectors be used. Hack to reuse the code without implementing a new DataType.
    #[arg(long = "use_minmax", default_value = "false")]
    pub use_minmax: bool,

    /// The dimension of the input dataset
    #[arg(long = "dim", required = true)]
    pub dim: usize,

    /// Optional value for the dimension of the dataset vectors when converted to full-precision vectors.
    /// This is only required when `use_minmax` = true.
    #[arg(long = "full_dim")]
    pub full_dim: Option<usize>,

    /// Distance function to use.
    #[arg(long = "dist_fn", default_value = "l2")]
    pub dist_fn: Metric,

    /// Path to the data file. The file should be in the format specified by the `data_type` argument.
    #[arg(long = "data_path", short, required = true)]
    pub data_path: String,

    /// Path to the index file. The index will be saved to this prefixed name.
    #[arg(long = "index_path_prefix", short, required = true)]
    pub index_path_prefix: String,

    /// Number of max out degree from a vertex.
    #[arg(long = "max_degree", short = 'R', default_value = "64")]
    pub max_degree: u32,

    /// Number of candidates to consider when building out edges
    #[arg(long = "l_build", short = 'L', default_value = "100")]
    pub l_build: u32,

    /// DRAM budget in GB for building the index
    #[arg(long = "build_DRAM_budget", short = 'M', default_value = "0")]
    pub build_dram_budget: f64,

    /// Quantization type to quantize the vector data.
    /// Supported values: 'FP', 'SQ_NBITS'(STDDEV=2.0 by default), 'SQ_NBITS_STDDEV'
    #[arg(long = "build_quantization_type", short, default_value = "FP")]
    pub build_quantization_type: QuantizationType,

    /// Number of PQ chunks to use.
    #[arg(long = "num_of_pq_chunks", required = true)]
    pub num_of_pq_chunks: usize,

    /// Number of threads to use.
    #[arg(long = "num_threads", short = 'T')]
    pub num_threads: Option<usize>,

    /// Set true for OPQ compression while using PQ distance comparisons for building the index, and false for PQ compression
    #[arg(long = "use_opq", short, default_value = "false")]
    pub use_opq: bool,
}
